### PR TITLE
feat: add structured cloning to Deno.core

### LINKED ISSUE #3557: Implement proper structured cloning between Workers
This is actually #3317, but we need a clearer issue name for tracking purposes.

Structured cloning implementation might be very useful in ways for transferring tasks between `Worker`s for load balancing, through low-cost transfer of `ArrayBuffer` (or even `SharedArrayBuffer`). For example, it might allow an HTTP request to move to another vacant thread if it somehow takes heavy CPU time to respond.

### PR DESCRIPTION
**Pull request by inteon**
First step towards #3557, as discussed in #9323. Also includes some simple tests.

### CODE REVIEW

File: core/serialize_deserialize.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  fn deserializer<'s>(
+    scope: &mut v8::HandleScope<'s>,
+    data: &[u8],
+    array_buffers: &'a mut ArrayBuffers,
+  ) -> v8::ValueDeserializer<'a, 's> {
+    v8::ValueDeserializer::new(scope, Box::new(Self { array_buffers }), data)
+  }
+}
```
Comment (bartlomieju) [inline core/serialize_deserialize.rs:26]: Seems like this struct is superfluous, why not use `v8::ValueSerializer` directly?

File: core/serialize_deserialize.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
```
Comment (bartlomieju) [inline core/serialize_deserialize.rs:1]: I believe we can put contents of this file into `bindings.rs` there's not that much code and we try to be frugal with files.

File: core/serialize_deserialize.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  buffer: &[u8],
+) -> Option<v8::Local<'s, v8::Value>> {
+  let mut array_buffers = ArrayBuffers::new(); // TODO(Inteon): fix array buffers
+  let mut value_deserializer =
+    SerializeDeserialize::deserializer(scope, &buffer, &mut array_buffers);
+  unsafe {
+    core::mem::transmute(
+      value_deserializer.read_value(scope.get_current_context()),
+    ) // TODO(Inteon): fix lifetime bug in rusty_v8?
+  }
```
Comment (bartlomieju) [inline core/serialize_deserialize.rs:93]: I'll try to get `rusty_v8` upgraded tonight

Comment (inteon) [reply core/serialize_deserialize.rs]: ok, fixed it

File: core/serialize_deserialize.rs (hunk 1 above)
Comment (inteon) [reply core/serialize_deserialize.rs]: I just pushed a simplified version.

File: core/serialize_deserialize.rs (hunk 2 above)
Comment (inteon) [reply core/serialize_deserialize.rs]: I just pushed a simplified version.

**Comment by inteon** [comment]
Blocked on #9466

**Review by bartlomieju — APPROVED** [review]
LGTM, thanks @inteon, great work! 

Before landing I'll wait for Ben's review

File: core/bindings.rs [hunk 4]
```diff
@@ -625,6 +627,90 @@ fn decode(
   };
 }
 
+struct SerializeDeserialize {}
+
+impl v8::ValueSerializerImpl for SerializeDeserialize {
+  #[allow(unused_variables)]
```
Comment (bnoordhuis) [inline core/bindings.rs:633]: What unused variables?

**Review by bnoordhuis — APPROVED** [review]
LGTM, thanks!

File: core/bindings.rs (hunk 4 above)
Comment (bartlomieju) [reply core/bindings.rs:633]: This function is required by `v8::ValueSerializerImpl` but is never used in `core/bindings.rs` (yet I think)

**Comment by ry** [comment]
Nice work @inteon - thanks!

**Pull request merged**